Expansion depth and recursive expansion

andruten opened the pull request:
Hi!

This may fix https://github.com/rsinger86/drf-flex-fields/issues/112 issue.

# Changes
Contributed two new settings to `FLEX_FIELDS_OPTIONS` dict. 
- `MAXIMUM_EXPANSION_DEPTH`, which is `Optional[int]`
- `RECURSIVE_EXPANSION_PERMITTED`, which is `bool`

- New for loop has been added in `_get_permitted_expands_from_query_param`, iterating through expansion fields looking for recursive expansions and expansion depth.
  - A ValidationError is raised when conditions are met in both cases. This exception can be overridden through `_recursive_expansion_found` and `_expansion_depth_exceeded` methods.
  - Expansion depth and recursion settings can be overridden as well through `maximum_expansion_depth` and `recursive_expansion_permitted` properties.

- Some tests have been added covering different scenarios
- Updated README
- added `venv` and `.idea` folders in gitignore.

rsinger86 commented:
This looks great, thank you. I just wanted to raise a few ideas for your consideration. If you want to keep it as is, I'm totally fine with that.

- Should we drop the `_` prefix on `_expansion_depth_exceeded` and `_recursive_expansion_found` since it appears the developer should be aware of them in case they want to customize?
- Is `_recursive_expansion_not_permitted` a better name for the `_recursive_expansion_found` method given that it's possible to allow recursive expansions?
- The recursive validation happens when the query param values are accessed, but it's possible to pass a list of expands directly to a serializer's ` __init__`, see [here](https://github.com/rsinger86/drf-flex-fields#serializer-options). Should we change it so the validation happens in both the query param and constructor scenarios?

andruten commented:
Than you for your comments @rsinger86. I've moved the validation to `_get_query_param_value` method, which applies in all three approaches. You can see the [commit](https://github.com/rsinger86/drf-flex-fields/pull/121/commits/0d3da72fd76009cf65cfde42ef81d1944da8845d). I've los dropped the `_` prefix and renamed the protected class attributes.

rsinger86 commented:
thanks!
